src/MatrixClient.ts

```diff
@@ -1880,10 +1885,62 @@ export class MatrixClient extends EventEmitter {
             headers["Authorization"] = `Bearer ${this.accessToken}`;
         }
-        return doHttpRequest(this.homeserverUrl, method, endpoint, qs, body, headers, timeout, raw, contentType, noEncoding);
+
+        const backoffTime = (nRetries: number) => {
+            return 1.6 ** nRetries * timeout;
+        }
+
+        const waitFor = (time: number) => {
+            return new Promise((resolve, reject) => setTimeout(_ => resolve(time), time));
+        };
+
+        // We would like to have the requestId from doHttpRequest so that logging retries can be a little more informative.
+        // The thing that stops us moving this code to http.ts is the configurable maxRetries, since if we have that here we can configure it per client.
+        // Another other reason for doing it here is giving the opportunity for a client to prevent retries for specific
+        // endpoints via the event emitter on the client, there could be better ways of doing this though?
+        let retries = 0;
+        let retry = true;
+        while (true) {
+            try {
+                return await doHttpRequest(this.homeserverUrl, method, endpoint, qs, body, headers, backoffTime(retries), raw, contentType, noEncoding);
+            } catch (err) {
+                const event = Object.freeze({
+                    retries,
+                    stopRetry() {
+                        LogService.debug('MatrixClient.ts', `Client request-${clientRequestId} stopped by a request.timeout listener.`)
+                        retry = false;
+                    },
+                    err
+                });
+                // FIXME: What about the rate limit errors? do we want to do something with those?
+                if (err.code === 'ESOCKETTIMEDOUT' || err.code === 'ETIMEDOUT') {
+                    LogService.info('MatrixClient.ts', `Retrying client request-${clientRequestId} ${method} ${endpoint}, current attempt ${retries}.`);
+                    this.emit('request.timeout', this.homeserverUrl, method, endpoint, event);
+                } else {
+                    throw err;
+                }
+
+                if (retries < this.maxRetries && retry) {
+                    await waitFor(backoffTime(retries));
+                    retries++;
+                } else if (this.maxRetries > 0) {
+                    LogService.error('MatrixClient.ts', `client request-${clientRequestId} retry limit reached ${this.maxRetries}.`)
+                    // FIXME: Do we want to wrap the error or something to explain how many times we've retried?
+                    throw err;
+                } else {
+                    // This means that they didn't want to retry either by intervention with a listener or by configuration.
+                    throw err;
+                }
+            }
+        }
     }
 }
 
+/**
+ * This is used to identify requests from the client that are being retried.
+ */
+let clientRequestId = 0;
```

Comment:
You're going to get race conditions reading this. You need to capture the value when you start `doRequest` if you wish to be consistent.